src/main/kotlin/io/github/zap/build/gradle/convention/ProjectExtension.kt

```diff
@@ -169,4 +171,99 @@ fun Project.configureShadow(destination: File,
     tasks.named("build", DefaultTask::class.java) {
         it.dependsOn(sj)
     }
+}
+
+fun Project.qs (name: String = "", options: ExternalModuleDependency.() -> Unit = {}) : ExternalModuleDependency.() -> Unit {
+    return {
+        val actualName = name.ifEmpty { this.name }
+        version {
+            if (isLocalTag(actualName)) {
+                it.require(getLocalVersion(actualName))
+            } else {
+                if(version != null && version!!.isNotEmpty()) {
+                    ext["qs-$actualName-ver"] = version
+                } else {
+                    val ver = getVerFromAncestor(actualName)
+                    if (ver != null) {
+                        it.require(ver)
+                    } else {
+                        logger.log(LogLevel.ERROR, "No version specified for dependency $actualName, fallback to 0.0.0-SNAPSHOT")
+                        it.require("0.0.0-SNAPSHOT")
+                    }
+                }
+            }
+        }
+        options()
+    }
+}
+
+@Suppress("UNCHECKED_CAST")
+internal fun Project.getQsLocals(): Set<String>? {
+    if(ext.has("qs-useLocals")) {
+        val cached = ext["qs-useLocals"]
+        if(cached is Set<*>) {
+            return cached as Set<String>
+        }
+    }
+
+    val useLocalClause = project.findProperty("qs.useLocal") as String?
+    if(useLocalClause == null) {
+        ext["qs-useLocals"] = null
+        return null
+    }
+
+    val useLocals = useLocalClause.split(",")
+        .map { it.trim() }
+        .filter { it.trim().isNotEmpty() }
+        .toSet()
+        ext["qs-useLocals"] = useLocals
+        return useLocals
+    }
+
+internal fun Project.isLocalTag(name: String): Boolean {
+    val locals = getQsLocals() ?: return false
+    if(locals.isEmpty()) return true
+
+    return locals.contains(name)
+}
+
+internal fun Project.getLocalVersion(name: String = ""): String {
+    if(name.isEmpty()) {
+        if(ext.has("qs-localVer")) {
+            val cached = ext["qs-localVer"]
+            if(cached is String) return cached
+        }
+
+        val localVersion = (project.findProperty("qs.localVer") as String? ?: System.getenv("QS_LOCAL_VER"))
+            ?.ifEmpty { "0.0.0-SNAPSHOT" } ?: "0.0.0-SNAPSHOT"
+
+        ext["qs-localVer"] = localVersion
+        return localVersion
+    } else { // I should refactor this later, kinda duplicate code
+        if(ext.has("qs-localVer-$name")) {
+            val cached = ext["qs-localVer-$name"]
+            if(cached is String) return cached
+        }
+
+        val localVersion = project.findProperty("qs.localVer.$name") as String?
```

Comment:
don't you want to check System env again like you did above?

Reply:
at this point there are env name are too long so I skip support it, I decided to only support the default

Reply:
we should either support System env everywhere or System env nowhere

Reply:
nah, I only give options to override in Gradle project props, system environment is kinda overkill